File: compiler/nodes/expressions/intrinsic.go
package expressions

import (
	"fmt"
	"strings"

	"wipple/codegen"
	"wipple/database"
	"wipple/syntax"
	"wipple/visit"
)

type IntrinsicExpressionNode struct {
	Name   string
	Inputs []database.Node
	Facts  *database.Facts
}

func (node *IntrinsicExpressionNode) GetFacts() *database.Facts {
	return node.Facts
}

func ParseIntrinsicExpression(parser *syntax.Parser) (*IntrinsicExpressionNode, *syntax.Error) {
	span := parser.Spanned()

	_, err := parser.Token("IntrinsicKeyword")
	if err != nil {
		return nil, err
	}

	parser.Commit("in this `intrinsic` expression")

	name, err := parser.Token("String")
	if err != nil {
		return nil, err
	}

	inputs, ok, err := syntax.ParseOptional(parser, func(parser *syntax.Parser) ([]syntax.Many[database.Node, struct{}], *syntax.Error) {
		return syntax.ParseMany(parser, 0, ParseAtomicExpression, syntax.ParseNothing)
	})
	if err != nil {
		return nil, err
	}

	var nodes []database.Node
	if ok {
		nodes = make([]database.Node, 0, len(inputs))
		for _, input := range inputs {
			nodes = append(nodes, input.Value)
		}
	}

	return &IntrinsicExpressionNode{
		Name:   name,
		Inputs: nodes,
		Facts:  database.NewFacts(span()),
	}, nil
}

func (node *IntrinsicExpressionNode) Visit(visitor *visit.Visitor) {
	visitExpression(visitor, node)

	for _, input := range node.Inputs {
		visitor.Visit(input)
	}
}

func (node *IntrinsicExpressionNode) Codegen(c *codegen.Codegen) error {
	span := database.GetSpanFact(node)

	name := strings.ReplaceAll(node.Name, "-", "_")
	c.WriteString(span, fmt.Sprintf("await __wipple_runtime_%s(", name))

	for _, input := range node.Inputs {
		if err := c.Write(input); err != nil {
			return err
		}

		c.WriteString(span, ", ")
	}

	c.WriteString(span, ")")

	return nil
}

File: compiler/src/nodes/expressions/intrinsic.rs
use crate::{
    codegen::{Codegen, CodegenCtx, CodegenResult, ir},
    database::{Node, NodeRef},
    nodes::{parse_atomic_expression, visit_expression},
    syntax::{ParseError, Parser, TokenKind},
    visit::{Visit, Visitor},
};

#[derive(Debug)]
pub struct IntrinsicExpressionNode {
    pub name: String,
    pub inputs: Vec<NodeRef>,
}

impl Node for IntrinsicExpressionNode {}

pub fn parse_intrinsic_expression(
    parser: &mut Parser<'_>,
) -> Result<IntrinsicExpressionNode, ParseError> {
    parser.token(TokenKind::IntrinsicKeyword)?;
    parser.commit("in this `intrinsic` expression");

    let name = parser.token(TokenKind::String)?;

    let inputs = if let Some(inputs) = parser.parse_optional(|parser| {
        parser.parse_many(0, parse_atomic_expression, |parser| parser.parse_nothing())
    })? {
        inputs.into_iter().map(|(node, _)| node).collect()
    } else {
        Vec::new()
    };

    Ok(IntrinsicExpressionNode { name, inputs })
}

impl Visit for IntrinsicExpressionNode {
    fn visit(&self, node: &NodeRef, visitor: &mut Visitor<'_>) {
        visit_expression(node, visitor);

        for input in &self.inputs {
            visitor.visit(input);
            visitor.edge(input, node, "input");
        }
    }
}

impl Codegen for IntrinsicExpressionNode {
    fn codegen(&self, node: &NodeRef, ctx: &mut CodegenCtx<'_>) -> CodegenResult {
        for input in &self.inputs {
            ctx.codegen(input)?;
        }

        ctx.instruction(ir::Instruction::Value {
            node: node.clone(),
            value: ir::Value::Runtime {
                name: self.name.clone(),
                inputs: self.inputs.clone(),
            },
        });

        Ok(())
    }
}
